### ISSUE TITLE
Supervised image classification using pretrained DinoV2

### ISSUE THREAD

**Issue by sahil02235** [ccharest93/LitDinov2_public#1]
Hi thanks for your code.
I wanted to know how to perform supervised classification tasks on dataset like cifar10, cifar100 and caltech101 with the pretrained dinov2. I just wanted to know how to add the classification head in this and maybe run for more 20 epochs just to see whether this finetuning is working fine or not.

Thanks,
Sahil

**Comment by woctezuma** [comment]
- https://github.com/facebookresearch/dinov2/issues/72#issuecomment-1529625079

**Comment by ccharest93** [comment]
Hi, thank you for your interest,
when making this repo i cut out a lot of the code needed for pretraining/fine-tuning. The point was to make it lean and understandable for inference on the pretrained weights trained by Meta. You are better suited using their original repo for your purpose, and the tutorial provided above might be a good starting point.

**Comment by itsprakhar** [comment]
The features extracted by DinoV2 are quite effective for downstream tasks and since the model weights they have provided are distilled to fit a small model they explicitly state that the gradients should not be propagated to the backbone (this will only degrade the performance of the model and hamper its capabilities to generalize). It's advised to extract features using the pre-trained model provided and train a small classification model for example a DNN.

Here's a repo you can refer to: https://github.com/itsprakhar/Simple-DinoV2-Classification

Here's some sample code that you may find useful

```import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader
from torchvision.datasets import ImageFolder
import torchvision.transforms as transforms

# Load the pre-trained feature extraction model
model = torch.hub.load('facebookresearch/dinov2', 'dinov2_vits14')

# Freeze the parameters of the pre-trained model
for param in model.parameters():
    param.requires_grad = False

# Define your own classification head
classifier = nn.Sequential(
    nn.Linear(384, num_classes),  # Modify `num_classes` according to your dataset
    nn.LogSoftmax(dim=1)
)

# Combine the feature extraction model with the classification head
model.classifier = classifier

# Define your training data and data transformations
train_dir = 'path/to/train/directory'
train_transform = transforms.Compose([
    transforms.Resize((224, 224)),  # Modify the size according to your requirements
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])  # Modify normalization values if needed
])

train_dataset = ImageFolder(train_dir, transform=train_transform)
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)

# Define your optimizer and loss function
optimizer = optim.SGD(model.parameters(), lr=learning_rate, momentum=0.9)
criterion = nn.NLLLoss()

# Train the classifier
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model.to(device)

for epoch in range(num_epochs):
    running_loss = 0.0
    for images, labels in train_loader:
        images = images.to(device)
        labels = labels.to(device)

        optimizer.zero_grad()

        # Extract features using the pre-trained model
        features = model(images)

        # Forward pass through the classification head
        outputs = model.classifier(features)

        # Calculate the loss
        loss = criterion(outputs, labels)

        # Backpropagation and optimization
        loss.backward()
        optimizer.step()

        running_loss += loss.item()

    # Print the average loss for the epoch
    epoch_loss = running_loss / len(train_loader)
    print(f'Epoch {epoch+1}/{num_epochs}, Loss: {epoch_loss:.4f}')

```

**Comment by sahil02235** [comment]
Thanks @itsprakhar for this. But I wanted this do for classification of xrays. I don't think only adding a classification head and freezing the backbone will help DinoV2 is not trained on medical images. It has been trained on lots of natural images so generalisation on natural images is expected to be good. I was thinking of freezing few intermediate layers and see the results.

**Comment by itsprakhar** [comment]
@sahil02235 As far as I understand, the primary advantage of using the DinoV2 model is the pre-trained weights learned by the new pretraining technique proposed in the paper. There's nothing new in DinoV2 in terms of the network architecture. So you can use the ViT or a variant of it if you want to finetune the deeper layers.

**Comment by ccharest93** [comment]
Even the pretraining technique isn't a contribution by the paper, they mostly demonstrated the scaling properties of that architecture/pretext task combination.
